File: templates/plugin_python/packer.py
"""A second module, showing the two declarations averager.py does not use.

A queueing input, because this module answers about a group of values rather than about the latest
one: a state input would drop everything that arrived between two iterations, and the frame would be
built from samples chosen by timing. And a blob output, which is the escape hatch of the C path — the
payload type set there is closed, so anything the platform does not name travels as bytes.

The struct import sits at module level on purpose: it is stdlib and costs nothing. A heavy dependency
belongs in initialize instead, because an import at module level runs when the bridge reads this file
and is paid by every pipeline, including those that never create this module.
"""

import struct

import atp

_FORMATS = {"little": "<{}d", "big": ">{}d"}


class Packer(atp.Module):
    name = "py_packer"
    version = (1, 0)

    value = atp.Input(atp.f64, queue=True, capacity=64)
    frame = atp.Output(atp.blob)
    packed = atp.Output(atp.i32)

    size = atp.Property(atp.i32, 4)
    byte_order = atp.Property(atp.text, "little", options=("little", "big"))

    def initialize(self):
        self.pending = []
        self.log("packing frames of {} values, {}-endian".format(self.size.get(), self.byte_order.get()))

    def iterate(self):
        value = self.value.take()
        if value is None:
            return atp.IDLE
        self.pending.append(value)
        if len(self.pending) < self.size.get():
            return atp.BUSY
        layout = _FORMATS[self.byte_order.get()].format(len(self.pending))
        self.frame.write(struct.pack(layout, *self.pending))
        self.packed.write(len(self.pending))
        self.pending = []
        return atp.BUSY

    def stop(self):
        self.pending = []

File: templates/plugin_lua/packer.lua
--- A module that turns a number into bytes, which is what a blob port is for.
---
--- Lua strings are byte strings, so a blob needs no encoding step and no dependency: string.pack
--- produces exactly the bytes the platform will carry as atp::io::blob.
local atp = require("atp")

local M = atp.module("lua_packer", { 1, 0 })

M.value = atp.input(atp.i32)

M.bytes = atp.output(atp.blob)

M.byte_order = atp.property(atp.text, "little", { options = { "little", "big" } })

function M:iterate()
    local v = self.value:take()
    if v == nil then
        return atp.IDLE
    end
    local format = self.byte_order:get() == "big" and ">i4" or "<i4"
    self.bytes:write(string.pack(format, v))
    return atp.BUSY
end

return M
